api/common/sharing/utils.go

```diff
@@ -12,64 +12,229 @@ import (
+	/* failures:
+	 * parameters with empty values
+	 * combination of ID with any other selector
+	 * parameters should not be empty, but the values of each parameter can be "" (empty string/object)
+	 */
+	if len(parameters) == 0 {
+		return util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	ID, IDExists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	name, nameExists := parameters[instance_sharing.ReferenceInstanceNameSelector]
+	plan, planExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	labels, labelsExists := parameters[instance_sharing.ReferenceLabelSelector]
+	var selectorLabels types.Labels
+	if err := util.BytesToObject([]byte(labels.Raw), &selectorLabels); labelsExists && err != nil {
+		return err
 	}
+	hasAtLeastOneSelector := selectorLabels != nil && len(selectorLabels) > 0 || nameExists && len(name.String()) > 0 || planExists && len(plan.String()) > 0
+	selectorsAndID := IDExists && len(ID.String()) > 0 && hasAtLeastOneSelector
+	emptyValues := len(ID.String()) == 0 && !hasAtLeastOneSelector
+	if selectorsAndID || emptyValues {
+		return util.HandleInstanceSharingError(util.ErrInvalidReferenceSelectors, "")
+	}
+	return nil
+}
+
+func getSharedInstances(ctx context.Context, repository storage.Repository, offeringID, tenantIdentifier, tenantID string) (types.ObjectList, error) {
+	queryParams := map[string]interface{}{
+		"tenant_identifier": tenantIdentifier,
+		"tenant_id":         tenantID,
+		"offering_id":       offeringID,
+	}
+	objectList, err := repository.QueryForList(ctx, types.ServiceInstanceType, storage.QueryForSharedInstances, queryParams)
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool, offeringID string) (*types.ServiceInstance, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
+
+	planSelectorVal, planSelectorExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	var selectorPlan *types.ServicePlan
+	if planSelectorExists && len(planSelectorVal.String()) > 0 {
+		selectorPlan, err = getSelectorPlan(ctx, repository, smaap, offeringID, planSelectorVal.String())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for i := 0; i < instances.Len(); i++ {
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists := parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists && len(selectorVal.String()) > 0 && !(instance.Name == selectorVal.String()) {
+			continue
+		}
+
+		if selectorPlan != nil && instance.ServicePlanID != selectorPlan.ID {
+			continue
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceLabelSelector]
+		if exists {
+			var selectorLabels types.Labels
+			if err := util.BytesToObject([]byte(selectorVal.Raw), &selectorLabels); err != nil {
+				return nil, err
```

Comment:
util.BytesToObject works without initializing selectorLabels?
if not, it's really strange that the tests are green

Reply:
It works, but I initialized it anyway.